### PR TITLE
Various fixes for msvc/cmake builds

### PR DESCRIPTION
**Pull request by phire**
There appeared to be quite a bit of bitrot

And I'm not sure if PCH support ever worked, I had to completely rewrite it to get it working.

### CODE REVIEW

File: Externals/libusb/CMakeLists.txt [hunk 1]
```diff
@@ -26,6 +26,8 @@ if(WIN32 OR CYGWIN)
       libusb/libusb/os/windows_common.c
       libusb/libusb/os/windows_usbdk.c
       libusb/libusb/os/windows_winusb.c
+      libusb/libusb/os/events_windows.c
+      libusb/libusb/os/threads_windows.c
```
Comment (shuffle2) [inline Externals/libusb/CMakeLists.txt:30]: already there

File: CMakeLists.txt [hunk 2]
```diff
@@ -272,6 +272,9 @@ if(CMAKE_C_COMPILER_ID MATCHES "MSVC")
     /wd5105 # macro expansion producing 'defined' has undefined behavior
   )
 
+  # Terminate building as soon as one error occurs
+  add_compile_options(/errorReport:prompt)
```
Comment (shuffle2) [inline CMakeLists.txt:276]: this is deprecated and relates to ICE (not normal compiler errors about source code) https://docs.microsoft.com/en-us/cpp/build/reference/errorreport-report-internal-compiler-errors?view=msvc-170

Comment (shuffle2) [reply CMakeLists.txt:276]: fwiw there is https://docs.microsoft.com/en-us/cpp/build/reference/diagnostics-compiler-diagnostic-options?view=msvc-170 which makes errors/warnings pretty-print in similar manner to clang (with `caret`)

Comment (phire) [reply CMakeLists.txt:276]: Interesting. 
It was one of the differences in the command line. cmake defaults it to queue, and visual studio defaults it to prompt. 

And I swear it made digging though backlog to find errors way easier.

Comment (phire) [reply CMakeLists.txt]: After restesting, it doesn't change anything, so I removed this.

**Comment by shuffle2** [comment]
@phire might need some tweaks since qt6 was merged

File: Source/CMakeLists.txt [hunk 3]
```diff
@@ -67,7 +67,13 @@ if (MSVC)
   # All files are encoded as UTF-8
   add_compile_options(/utf-8)
 
-  # Use PCH
+  # Ignore warnings in external headers
+  add_compile_options(/external:anglebrackets)
+  add_compile_options(/external:W0)
+  
+  # TODO: Enable this once PR #10559 is merged
```
Comment (Tilka) [inline Source/CMakeLists.txt:74]: That PR is now merged.

File: Source/Core/DolphinQt/CMakeLists.txt [hunk 4]
```diff
@@ -371,6 +371,11 @@ if (WIN32)
 endif()
 
 if (MSVC)
+  # Disable some warnings
+  target_compile_options(dolphin-emu PRIVATE
+    /wd5054 # operator '+': deprecated between enumerations of different types (types defined in Qt headers)
```
Comment (Tilka) [inline Source/Core/DolphinQt/CMakeLists.txt:376]: Why is this needed when you already disable all warnings in external code?

Comment (shuffle2) [reply Source/Core/DolphinQt/CMakeLists.txt:376]: Fwiw shouldn’t be required anymore as qt fixed their code

**Comment by shuffle2** [comment]
@phire fwiw #10652 removes the need to fiddle with ignoring warnings generated from `fmt`

**Comment by shuffle2** [comment]
I did some profiling and I think some more easy-ish big speed boost opportunities are:
* pch for C files (for externals)
* pch for DolphinQt (including moc outputs) which includes Qt headers

**Comment by AdmiralCurtiss** [comment]
FYI to get this to build on VS2022 with Qt6 I had to add a `CMAKE_PREFIX_PATH` pointing at the Qt directory, as else it wouldn't find Qt6 (even if I set `Qt6_DIR`). So in `CMakeSettings.json` add:

~~~
        {
          "name": "CMAKE_PREFIX_PATH",
          "value": "${workspaceRoot}\\Externals\\Qt\\Qt6.3.0\\x64"
        },
~~~

for each config -- `ARM64` instead of `x64` for the ARM builds, presumably.

Additionally, I had to make an empty `mkspecs\win32-msvc` in `Externals\Qt\Qt6.3.0\x64`, as apparently the Qt CMake files set that as an include directory? We probably should add that in the Qt repository.

**Comment by AdmiralCurtiss** [comment]
The `--no-angle` also needs to be removed in the Qt deploy step, as that option no longer exists in Qt6.

**Comment by shuffle2** [comment]
> FYI to get this to build on VS2022 with Qt6 I had to add a `CMAKE_PREFIX_PATH` pointing at the Qt directory, as else it wouldn't find Qt6 (even if I set `Qt6_DIR`). So in `CMakeSettings.json` add:
> 
> ```
>         {
>           "name": "CMAKE_PREFIX_PATH",
>           "value": "${workspaceRoot}\\Externals\\Qt\\Qt6.3.0\\x64"
>         },
> ```
> 
> for each config -- `ARM64` instead of `x64` for the ARM builds, presumably.

You tried this PR rebased on master? But, setting it in CMAKE_PREFIX_PATH instead of X_DIR is fine (actually what Qt docs recommends), anyway.

> Additionally, I had to make an empty `mkspecs\win32-msvc` in `Externals\Qt\Qt6.3.0\x64`, as apparently the Qt CMake files set that as an include directory? We probably should add that in the Qt repository.

Do you know what it's looking for in there? Most of the files are for qmake it seems, except for `mkspecs/preliminary/<BuildType>/qmodule.cmake`, which are empty files... Maybe we just need to tell Qt not to bother with qmake stuff somehow?

**Comment by AdmiralCurtiss** [comment]
Yes, this PR rebased on master.

And no clue what files it wants from there, it seems to build fine even with none of them. `qplatformdefs.h` seems like a reasonable guess though.

**Comment by AdmiralCurtiss** [comment]
85296d1920d79d63cde3fe531fe44a501c24a8d3 if you want to cherry-pick the Qt6 fixes.

**Comment by shuffle2** [comment]
you can probably get rid of a lot of other stuff from the qt deployment? (take a look at what msbuild build copies) maybe files that don't exist (in our Qt submodule) get silently ignored, but probably better to have it actually correct...

File: CMakeLists.txt [hunk 5]
```diff
@@ -9,15 +9,30 @@ cmake_minimum_required(VERSION 3.10)
...
+if(POLICY CMP0091)
+  cmake_policy(SET CMP0091 NEW) # MSVC runtime library flags are selected by an abstraction.
+  cmake_policy(SET CMP0092 NEW) # MSVC warning flags are not in CMAKE_{C,CXX}_FLAGS by default.
+endif()
+
 project(dolphin-emu)
 
+if (MSVC)
+  set(CMAKE_C_STANDARD 99)
+  set(CMAKE_CXX_STANDARD 20)
```
Comment (shuffle2) [inline CMakeLists.txt:30]: https://stackoverflow.com/a/64890021 seems the way to do this is to set CXX to 23 and remove the explicit "latest" option?

Comment (AdmiralCurtiss) [reply CMakeLists.txt:30]: For what it's worth as written it seems to work, I get `-std:c++latest` in the build command line.

Comment (shuffle2) [reply CMakeLists.txt:30]: fwiw `C:\Program Files\Microsoft Visual Studio\2022\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\share\cmake-3.22\Modules\Compiler\MSVC-CXX.cmake` has this:
```cmake
  if(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 19.29.30129)
    set(CMAKE_CXX20_STANDARD_COMPILE_OPTION "-std:c++20")
    set(CMAKE_CXX20_EXTENSION_COMPILE_OPTION "-std:c++20")
    set(CMAKE_CXX23_STANDARD_COMPILE_OPTION "-std:c++latest")
    set(CMAKE_CXX23_EXTENSION_COMPILE_OPTION "-std:c++latest")
  elseif(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 19.12.25835)
    set(CMAKE_CXX20_STANDARD_COMPILE_OPTION "-std:c++latest")
    set(CMAKE_CXX20_EXTENSION_COMPILE_OPTION "-std:c++latest")
  endif()
```
which is i guess what causes just `set(CMAKE_CXX_STANDARD 23)` to enable `latest` on current version (and it will probably break in the future, great! :D)

**Comment by AdmiralCurtiss** [comment]
@shuffle2 I think I'm done for now, do you want to review?

File: CMakeLists.txt [hunk 6]
```diff
@@ -9,15 +9,29 @@ cmake_minimum_required(VERSION 3.10)
...
+  cmake_policy(SET CMP0092 NEW) # MSVC warning flags are not in CMAKE_{C,CXX}_FLAGS by default.
+endif()
+
 project(dolphin-emu)
 
+if (MSVC)
+  set(CMAKE_C_STANDARD 99)
+  set(CMAKE_CXX_STANDARD 23)
+  set(CMAKE_CXX_STANDARD_REQUIRED ON)
+  set(CMAKE_CXX_EXTENSIONS OFF)
```
Comment (shuffle2) [inline CMakeLists.txt:32]: afaiui, this causes cmake to pass `Za` for C++ sources(?), something that msbuild setup is currently not doing. It's probably better to leave out for now

File: CMakeLists.txt [hunk 7]
```diff
@@ -9,15 +9,29 @@ cmake_minimum_required(VERSION 3.10)
...
+# This is done in the universal binary building script to build a binary that
 # runs on 10.13 on x86_64 computers, while still containing an arm64 slice.
 
 set(CMAKE_OSX_DEPLOYMENT_TARGET "10.14.0" CACHE STRING "")
 
 set(CMAKE_USER_MAKE_RULES_OVERRIDE "CMake/FlagsOverride.cmake")
 
+# Optionally enable these polcies so older versions of cmake don't break.
+# we only need them for MSVC
+if(POLICY CMP0091)
```
Comment (shuffle2) [inline CMakeLists.txt:21]: setting `CMP0117` to `NEW` should make the `GR` stuff more sane to setup. (or maybe just `cmake_policy(VERSION ${CMAKE_VERSION})` to advance all behavior on cmake/msvc?)

File: CMakeLists.txt [hunk 8]
```diff
@@ -244,16 +258,70 @@ elseif(CMAKE_GENERATOR MATCHES "Visual Studio")
   add_compile_options("/MP")
 endif()
 
-if(CMAKE_C_COMPILER_ID MATCHES "MSVC")
+if(MSVC)
+  if(POLICY CMP0091)
```
Comment (shuffle2) [inline CMakeLists.txt:262]: I think the check should be against cmake version instead (3.22 aligns with our current VS requirements, 3.20 is required for CMP0117 )

File: Source/Core/DolphinQt/CMakeLists.txt [hunk 9]
```diff
@@ -394,6 +391,13 @@ if (MSVC)
   target_compile_options(dolphin-emu PRIVATE "${qtGui}")
   target_compile_options(dolphin-emu PRIVATE "${qtGuiPriv}")
   target_compile_options(dolphin-emu PRIVATE "${qtWidgets}")
+
+  if ("${QT_VERSION_MAJOR}" GREATER_EQUAL 6)
```
Comment (shuffle2) [inline Source/Core/DolphinQt/CMakeLists.txt:395]: see comment about `CMP0117`

File: CMakeLists.txt [hunk 10]
```diff
@@ -244,16 +258,70 @@ elseif(CMAKE_GENERATOR MATCHES "Visual Studio")
...
+    # We really kind of want this policy
+    message(FATAL_ERROR "please update cmake to at least 3.15")
+  endif()
+
   check_and_add_flag(EXCEPTIONS /EHsc)
   dolphin_compile_definitions(_DEBUG DEBUG_ONLY)
 
+  # Disable RTTI
+  # Unfortunately /GR is in the default compile flags for MSVC so we have to find and replace it.
+  foreach (flag CMAKE_CXX_FLAGS CMAKE_CXX_FLAGS_DEBUG CMAKE_CXX_FLAGS_RELEASE CMAKE_CXX_FLAGS_MINSIZEREL CMAKE_CXX_FLAGS_RELWITHDEBINFO)
```
Comment (shuffle2) [inline CMakeLists.txt:274]: see comment about `CMP0117`

File: CMakeLists.txt (hunk 6 above)
Comment (AdmiralCurtiss) [reply CMakeLists.txt:32]: As far as I can tell, this option does absolutely nothing on MSVC.

File: CMakeLists.txt (hunk 8 above)
Comment (AdmiralCurtiss) [reply CMakeLists.txt]: I'm not sure what advantage this has. By saying `if (POLICY CMP0117)` we state exactly what we care about -- the policy existing -- and not some arbitrary CMake version that could mean anything.

**Comment by AdmiralCurtiss** [comment]
@shuffle2 One more check please? I'd like to get this merged so this actually builds on master.

File: CMakeLists.txt (hunk 6 above)
Comment (shuffle2) [reply CMakeLists.txt:39]: Remove it then?

File: CMakeLists.txt (hunk 8 above)
Comment (shuffle2) [reply CMakeLists.txt]: I said this because I’d prefer 
cmake_policy(VERSION ${CMAKE_VERSION}) 
but if that’s not done then checking against highest policy we manually enable is fine I guess

Comment (AdmiralCurtiss) [reply CMakeLists.txt]: As in, always set all policies supported on whatever the current CMake versions is on? That seems dangerous and very much invites the behavior changing invisibly when you upgrade CMake, which is kinda what policies are designed to avoid. I'd be okay with `cmake_policy(VERSION 3.22)` (which is the version currently used by VS), though, after testing it works with our files anyway.

Comment (shuffle2) [reply CMakeLists.txt]: yes - the thing is that for msvc we know/expect cmake to be the one that gets installed by VS installer. So it's not a random version. I think cmake "getting out of the way" and behaving natively as the one that is installed with msvc is more intuitive and requires less code, but explicitly specifying the version is fine, too (we just have to bump it whenever we bump msvc min version).

anyway, how it is now is fine.

**Comment by AdmiralCurtiss** [comment]
@dolphin-emu-bot rebuild
